### ISSUE TITLE
New data structure for presentation of edges

### ISSUE THREAD

**Issue by laa** [orientechnologies/orientdb#4491]
I thought in background how to find quick way to iterate over/add edges without slowdown on removal of edges.

So following complex data structure came into my mind:
RidBag interface is used as base for presentaton of edges.
So main methods of ridbag are add/remove/iterate (obviously).

By RidBag's nature implementation of add/iterate is very simple we may use linked list (which in reality would be presented as "skip list") and have super fast performance.

So what about fast implementation of remove method ?
Yesterday I watched ideas which are put behind cola tree (super cool thing BTW but not all secrets of this data structure were made public of course) and decided to borrow some.

So let's take into consideration that remove operation is not really frequent operation. If we pay attention on that we may implement this task using additional data structures.
We may use hash map (implemented like linear hashing hash map) and counting bloom filter.
Every time when we remove item from ridbag we put it in hash map and update counting bloom filter.
During iteration/add we check if item is in bloom filer (and then in hash map of course), if it is we skip it, otherwise we publish/add this item to the user.

To prevent accumulation of garbage we will have background thread (cola tree has the same one and everybody is happy about it) which merge deleted items into skip list.
WDYT about it ?

Seems like it will bring high performance to the system during massive insertion and iteration over edges.

**Comment by lvca** [comment]
If I understood correctly, you would use bloom filter only for delete stuff, right? So if you remove an edge, let's say a RID, you could put in the hashmap + bloom filter, but how do you know if the edge is really contained in the list if you don't traverse it all? Wouldn't be much better to use the counting Bloom Filter to store contained RIDs so you can query it to see if a RID is contained? This would also speedup lookups.

**Comment by laa** [comment]
Missed your comment at the morning. When you remove item from collection does not matter whether it is contained inside collection or not. We should get final result that is item for sure is absent in collection after remove. That is what is solved by this data structure.  You see issue with bloom filter in list is following if you remove item you in 90% of cases remove item which exists in collection and because bloom filter is used to detect absence of item in  data structure it means that you will do full scan in 90% of cases so approach with bloom filter on list does not add any value and will be even slower then simple full scan because we add bloom filter overhead.

About you question : " but how do you know if the edge is really contained in the list if you don't traverse it all?" . We merely do not need this knowledge to perform item removal.

**Comment by lvca** [comment]
so you could invoke a 1M of remove of random RIDs, and all the RID would be saved+bloomfiltered in the collection before the clean, right?

**Comment by laa** [comment]
they will be save in linear hashing table and added to bloom filter.

**Comment by lvca** [comment]
Forgot to say that I like this. We should also be compatible with the past, so what's your proposal and ETA to do that?

**Comment by laa** [comment]
@lvca I may give only pretty rough ETA. 
For linear hashing 7 days for implementation and testing is enough. For counting bloom filter and it's integration I think 7 days is enough. I think skip list implementation will take about 4 days and clean up Clean up thread implementation will take about 3 - 5 days (it sounds as simple think but in reality it is not :-) ) 

There is one note we introduced contains method (which I did not support :-) ) which will become slow after this change , so we need either rewrite our intersect function or make contains faster. But IMHO that is subject of another issue because current one is very important even without contains method.
